gobblin-runtime/src/main/java/gobblin/runtime/FsDatasetStateStore.java

```diff
@@ -0,0 +1,226 @@
+   *
+   * @param jobName the job name
+   * @return a {@link Map} from dataset URNs to the latest {@link JobState.DatasetState}s
+   * @throws IOException if there's something wrong reading the {@link JobState.DatasetState}s
+   */
+  public Map<String, JobState.DatasetState> getLatestDatasetStatesByUrns(String jobName) throws IOException {
+    Path stateStorePath = new Path(this.storeRootDir, jobName);
+    if (!this.fs.exists(stateStorePath)) {
+      return ImmutableMap.of();
+    }
+
+    FileStatus[] stateStoreFileStatuses = this.fs.listStatus(stateStorePath, new PathFilter() {
+      @Override
+      public boolean accept(Path path) {
+        return path.getName().endsWith(CURRENT_DATASET_STATE_FILE_SUFFIX + DATASET_STATE_STORE_TABLE_SUFFIX);
+      }
+    });
+
+    if (stateStoreFileStatuses == null || stateStoreFileStatuses.length == 0) {
+      return ImmutableMap.of();
+    }
+
+    Map<String, JobState.DatasetState> datasetStatesByUrns = Maps.newHashMap();
+    for (FileStatus stateStoreFileStatus : stateStoreFileStatuses) {
+      List<JobState.DatasetState> previousDatasetStates = getAll(jobName, stateStoreFileStatus.getPath().getName());
+      if (!previousDatasetStates.isEmpty()) {
+        // There should be a single dataset state on the list if the list is not empty
+        JobState.DatasetState previousDatasetState = previousDatasetStates.get(0);
+        datasetStatesByUrns.put(previousDatasetState.getProp(ConfigurationKeys.DATASET_URN_KEY,
+            ConfigurationKeys.DEFAULT_DATASET_URN), previousDatasetState);
+      }
+    }
+
+    // The dataset (job) state from the deprecated "current.jst" will be read even though
+    // the job has transitioned to the new dataset-based mechanism
+    if (datasetStatesByUrns.size() > 1) {
+      datasetStatesByUrns.remove(ConfigurationKeys.DEFAULT_DATASET_URN);
+    }
+
+    return datasetStatesByUrns;
+  }
+
+  public JobState.DatasetState getLatestDatasetState(String storeName, String datasetUrn) throws IOException {
+    String alias = Strings.isNullOrEmpty(datasetUrn) ?
+        CURRENT_DATASET_STATE_FILE_SUFFIX + DATASET_STATE_STORE_TABLE_SUFFIX
+        : datasetUrn + "-" + CURRENT_DATASET_STATE_FILE_SUFFIX + DATASET_STATE_STORE_TABLE_SUFFIX;
+    return get(storeName, alias, datasetUrn);
+  }
+
+  /**
+   * Persist a given {@link JobState.DatasetState}.
+   *
+   * @param datasetUrn the dataset URN
+   * @param datasetState the {@link JobState.DatasetState} to persist
+   * @throws IOException if there's something wrong persisting the {@link JobState.DatasetState}
+   */
+  public void persistDatasetState(String datasetUrn, JobState.DatasetState datasetState) throws IOException {
+    String jobName = datasetState.getJobName();
+    String jobId = datasetState.getJobId();
+
+    String tableName = Strings.isNullOrEmpty(datasetUrn) ? jobId + DATASET_STATE_STORE_TABLE_SUFFIX
+        : datasetUrn + "-" + jobId + DATASET_STATE_STORE_TABLE_SUFFIX;
```

Comment:
We need some escaping here as the URNs may contain ':' and other characters which are not allowed. Overall, I think we have to enclose this logic in a separate function for easier managibility.